tm2/pkg/crypto/keys/client/maketx.go

```diff
@@ -234,3 +269,127 @@ func ExecSignAndBroadcast(
+		}
+		tx.Fee = std.NewFee(cfg.GasWanted, gasfee)
+		return nil
+	}
+
+	// Auto gas estimation
+	// Get the remote client
+	remote := cfg.RootCfg.Remote
+	if remote == "" {
+		return errors.New("missing remote url for gas estimation")
+	}
+
+	cli, err := client.NewHTTPClient(remote)
+	if err != nil {
+		return errors.Wrap(err, "creating HTTP client for gas estimation")
+	}
+
+	// Set temporary fee for simulation
+	// We need to set a non-zero fee for the simulation to work
+	tempGasFee := std.NewCoin("ugnot", 1000000) // 1 GNOT temporary fee
+	tempGasWanted := int64(10000000)            // 10M gas temporary limit
+
+	// Store original fee and signatures to restore if needed
+	originalFee := tx.Fee
+	originalSignatures := tx.Signatures
+
+	// Set temporary values for simulation
+	tx.Fee = std.NewFee(tempGasWanted, tempGasFee)
+
+	// Add a dummy signature for simulation to pass signature validation
+	// We need at least one signature with the correct length for the ante handler
+	if len(tx.Signatures) == 0 {
+		dummySig := std.Signature{
+			PubKey:    nil,              // Can be nil for simulation
+			Signature: make([]byte, 64), // Empty signature, correct length
+		}
+		tx.Signatures = []std.Signature{dummySig}
+	}
+
+	// Serialize the transaction for simulation
+	bz, err := amino.Marshal(tx)
+	if err != nil {
+		tx.Fee = originalFee // Restore original
+		tx.Signatures = originalSignatures
+		return errors.Wrap(err, "marshaling tx for gas estimation")
+	}
+
+	// Simulate the transaction to get gas usage
+	res, err := SimulateTx(cli, bz)
+	if err != nil {
+		tx.Fee = originalFee // Restore original
+		tx.Signatures = originalSignatures
+		return errors.Wrap(err, "simulating transaction for gas estimation")
+	}
+
+	if res.CheckTx.IsErr() {
+		tx.Fee = originalFee // Restore original
+		tx.Signatures = originalSignatures
+		return errors.Wrapf(res.CheckTx.Error, "transaction check failed during gas estimation: %s", res.CheckTx.Log)
+	}
+
+	if res.DeliverTx.IsErr() {
+		tx.Fee = originalFee // Restore original
+		tx.Signatures = originalSignatures
+		return errors.Wrapf(res.DeliverTx.Error, "transaction delivery failed during gas estimation: %s", res.DeliverTx.Log)
+	}
+
+	// Get the estimated gas used and add a buffer
+	gasUsed := res.DeliverTx.GasUsed
+	// Add 10% buffer to the estimated gas
+	gasWanted := gasUsed + (gasUsed / 10)
+
+	// Query gas price if fee is not set
+	var gasFee std.Coin
+	if cfg.GasFee == "" {
+		gp := std.GasPrice{}
+		qres, err := cli.ABCIQuery("auth/gasprice", []byte{})
+		if err != nil {
+			return errors.Wrap(err, "querying gas price for fee estimation")
+		}
+		err = amino.UnmarshalJSON(qres.Response.Data, &gp)
+		if err != nil {
+			return errors.Wrap(err, "unmarshaling gas price result")
+		}
+
+		if gp.Gas == 0 {
+			// No gas price set, use a default fee
+			gasFee = std.NewCoin("ugnot", 1000000) // Default fee
+		} else {
+			// Calculate fee based on gas price
+			feeAmount := gasWanted/gp.Gas + 1
```

Comment:
The gas fee calculation should use [the same math](https://github.com/gnolang/gno/blob/de4b5b56c60126373ec0702234c196fdae365a0b/tm2/pkg/crypto/keys/client/broadcast.go#L157-L161) with `overflow.Mulp` as the gnokey CLI `-simulate only`: